content/10-managing-databases/07-introduction-database-caching.mdx

```diff
@@ -0,0 +1,97 @@
+
+Alternatively, the main drawback of a cache-aside strategy is the window being open for inconsistency from the database. Generally, any data being written will go to the database directly. Therefore, the cache may have a period of inconsistency with the primary database. There are different cache strategies to combat this depending on your needs.
+
+### Read-through
+
+In a read through cache arrangement, the cache sits between the application and the database. It can be envisioned like a straight line from application to database with the cache in the middle. In this strategy, the application will always speak with the cache for a read, and when there is a cache hit, the data is immediately returned. In the case of a cache miss, the cache will populate the missing data from the database and then return it to the application. For any data writes, the application will still go directly to the database.
+
+Read-through caches are also good for read-heavy workloads. The main differences between read-through and cache-aside is that in a cache-aside strategy the application is responsible for fetching the data and populating the cache, while in a read-through setup, the logic is done by a library or some separate cache provider. A read-through setup is similar to a cache-aside in regards to potential data inconsistency between cache and database.
+
+A read-through caching strategy also has the disadvantage of needing to go to the database to get the data anytime a new read request comes through. This data has never been cached before so therefore the data needs to be loaded. It is common for developers to mitigate this delay by ‘warming’ the cache by issuing likely to happen queries manually.
+
+### Write-through
+
+A write-through caching strategy differs from the previously two mentioned because instead of writing data to the database, it will write to the cache first and the cache immediately writes to the database. The arrangement can still be visualized like the read-through strategy, in a straight line with the application in the middle.
+
+The benefit to a write-through strategy is that the cache is ensured to have any written data and no new read will experience delay while the cache requests it from the main database. If solely making this arrangement, there is the big disadvantage of extra write latency because the action must go to the cache and then to the database. 
+
+The real benefit comes from pairing a write-through with a read-through cache. This strategy will adopt all the aforementioned benefits of the read-through caching strategy with the added benefit of removing the potential for data inconsistency. 
+
```

Comment:
I think it's worth a passing mention that there's a big risk here if the write returns as successful before the database has been updated (if there's a DB failure before the data can be securely forwarded from the cache to the DB).

Reply:
After some further reading I did out of curiosity, I might be off here ^.  I think I misunderstood this strategy a bit.  So write-through seems to be:

1. application writes change to cache
2. application (or cache) writes changes to main DB
3. the write returns as successful

I think the part that wasn't clear to me initially was that the DB is updated at the same time as the cache with this strategy (no delay, no successful write until both sources are updated).  That means that the danger of data loss isn't a risk with this strategy.  Sorry for the confusion.  I'm not sure if my misunderstanding was just me making assumptions here or not, but thought I'd mention it in case it's helpful with the explanation.

Reply:
Yeah the implication here is that the cache writes immediately to the database. I still feel like there is a bit of risk though because immediately does not mean simultaneously so there is some sort of delay probably in the fractions of milliseconds.